keps/sig-network/3015-prefer-same-node/README.md

```diff
@@ -0,0 +1,496 @@
+because it's faster, or because it's cheaper?)
+
+However, we are not likely to ever end up with a highly-detailed
+system for explaining the user's relative preferences for latency,
+bandwidth, cost, server load, etc, and many users have indicated that
+they would be very happy with a way to _just_ express "prefer same
+zone" without worrying that this might thwart some theoretical future
+proxy. (And of course, if it _does_ thwart some theoretical future
+proxy, they could just change the Service definition at that point.)
+
+Also, while these alternative traffic distribution modes risk
+overloading endpoints if used poorly, the API does not need to assume
+that the service proxy is the _only_ component in the system that
+cares about balancing endpoint load; the user may be using other APIs
+such as Pod Topology Spread Constraints or Horizontal Autoscaling to
+ensure that load is handled appropriately, and in those scenarios, the
+alternatively traffic distribution modes may be completely safe.
+
+## Design Details
+
+### Renaming/Deprecation of `PreferClose`
+
+KEP-4444 defines `PreferClose` as:
+
+> * `PreferClose`: Indicates a preference for routing traffic to endpoints that
+>   are topologically proximate to the client. The interpretation of
+>   "topologically proximate" may vary across implementations and could encompass
+>   endpoints within the same node, rack, zone, or even region.
+
+We will add a new `TrafficDistribution` value, `PreferSameZone`, which
+is defined as follows:
+
+* `PreferSameZone`: Indicates a preference for routing traffic to
+  endpoints that are in the same zone as the client. In general, the
+  proxy should always route to a same-zone endpoint if any is
+  available.
+
+And we will say that henceforth, `PreferClose` is just a deprecated
+alias for `PreferSameZone`, and never means anything other than
+"prefer same zone".
+
+Note that this does not change the existing implementation of
+`PreferClose` at all, it only removes the possibility of future
+changes.
+
+(In theory, service proxies other than kube-proxy may already have
+been implementing `PreferClose` with semantics other than the ones
+kube-proxy used. Those implementations will need to be updated when
+they add support for this KEP.)
+
+### Addition of `PreferSameNode`
+
+We will add a new `TrafficDistribution` value, `PreferSameNode`, which
+is defined as follows:
+
+* `PreferSameNode`: Indicates a preference for routing traffic to
+  endpoints that are on the same node as the client. In general, the
+  proxy should always route to a same-node endpoint if any is
+  available.
+
+We will add a new field to `discoveryv1.EndpointHints`:
+
+```golang
+// EndpointHints provides hints describing how an endpoint should be consumed.
+type EndpointHints struct {
+        ...
+
+	// forNodes indicates the node(s) this endpoint should be targeted by.
+	// +listType=atomic
+	ForNodes []string `json:"forNodes,omitempty" protobuf:"bytes,2,name=forNodes"`
```

Comment:
Is the ForNodes field just a slice of strings, whereas the ForZones field is defined as a [slice of ForZone](https://github.com/kubernetes/kubernetes/blob/v1.32.1/pkg/apis/discovery/types.go#L134-L145)?

Reply:
The `ForZone` struct was apparently added to support future ideas like having multiple zones with `weight` values... (https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/2433-topology-aware-hints/README.md#future-api-expansion)

I guess we could do the same for `ForNodes`... ? I think the idea of weighted hints is based more on the "smart controllers / ambiguous API" model of TopologyAwareHints, while TrafficDistribution has moved more toward "simple controllers / explicit API" so that sort of future API expansion seems unlikely.

Reply:
As mentioned in https://github.com/kubernetes/enhancements/pull/5140#discussion_r1948103197, even if we are in the "smart controllers / ambiguous API" model, actually, there's no need to include weight information in ForZones hints. It’s enough to determine the weight within the service proxy implementation, so it's unnecessary.

However, looking at this wording, it might be difficult to definitively say that a string slice is the best choice...🤷‍♂️ For example, if fields other than `trafficDistribution` are added, allowing users to specify the behavior in more detail, is there a possibility that Hints would need to include additional information?
> Although it is unclear if either expansion will be necessary, the API is designed in a way to make expansions straightforward.